File: go/day01.go
package main

import (
	"math"
	"sort"
	"strconv"
	"strings"
)

func Day01(input string) (int, int) {
	lines := strings.Split(input, "\n")
	list1 := make([]int, len(lines))
	list2 := make([]int, len(lines))
	for _, line := range lines {
		nums := strings.Fields(line)
		a, err1 := strconv.Atoi(nums[0])
		b, err2 := strconv.Atoi(nums[1])

		if err1 != nil || err2 != nil {
			continue
		}

		list1 = append(list1, a)
		list2 = append(list2, b)
	}

	sort.Slice(list1, func(i, j int) bool {
		return list1[i] < list1[j]
	})

	sort.Slice(list2, func(i, j int) bool {
		return list2[i] < list2[j]
	})

	sum := 0

	for i := range list1 {
		sum += int(math.Abs(float64(list1[i] - list2[i])))
	}

	sim := 0

	j := 0

	for _, num := range list1 {
		for j < len(list2) && list2[j] < num {
			j++
		}

		for j < len(list2) && list2[j] == num {
			sim += num
			j++
		}
	}

	return sum, sim
}

File: rust/src/day01.rs
pub fn run(input: &str) -> (i32, i32) {
    // let it = input
    //     .lines()
    //     .map(|line| {
    //         line.trim()
    //             .split_ascii_whitespace()
    //             .map(|num| num.parse::<i32>().unwrap())
    //             .collect::<Vec<i32>>()
    //     })
    //     .collect::<Vec<Vec<i32>>>();

    // let mut list1 = it.iter().map(|x| x[0]).collect::<Vec<i32>>();
    // let mut list2 = it.iter().map(|x| x[1]).collect::<Vec<i32>>();

    let (mut list1, mut list2): (Vec<i32>, Vec<i32>) = input
        .lines()
        .map(|line| {
            let mut nums = line
                .trim()
                .split_ascii_whitespace()
                .map(|num| num.parse::<i32>().unwrap());
            (nums.next().unwrap(), nums.next().unwrap())
        })
        .unzip();

    list1.sort();
    list2.sort();

    let mut sum = 0;

    for i in 0..list1.len() {
        sum += (list1[i] - list2[i]).abs();
    }

    let mut sim = 0;

    let mut j = 0;

    for num in list1.iter() {
        while j < list2.len() && list2[j] < *num {
            j += 1;
        }

        while j < list2.len() && list2[j] == *num {
            sim += num;
            j += 1;
        }
    }

    // for num in list1.iter() {
    //     for num2 in list2.iter() {
    //         if num == num2 {
    //             sim += num;
    //         }
    //     }
    // }

    (sum, sim)
}
